File: Easy/Two_Sum/Two_Sum.js
// Runtime: 52 ms, faster than 94.04% of JavaScript online submissions for Two Sum.
// Memory Usage: 35.8 MB, less than 16.12% of JavaScript online submissions for Two Sum.
// 02/24/2020
// MM/DD/YY

function shell_sort(array) {
  let gap = Math.floor(array.length/2)
  while (gap >= 1) {
      //console.log(array, gap)
      for (i = gap; i < array.length; i ++) { //iterate through array, starting from gap
          let comparator = array[i] //make comparisons with this
          let index //in case of negative index
          let output //for accessing x outside the array

          for (x = i; x > gap-2; x -= gap) { //iterate throguh array with gap as the step
              output = x //for accessing x outside the array
              if (x-gap < 0) { //in case of negative index
                  index = array.length-x-gap
              } else {
                  index = x-gap
              }
              
              if (array[index] <= comparator) { //break when correct spot is found
                  break 
              } else { //otherwise, move elements forward to make space
                  array[x] = array[index] 
              }
          }
          array[output] = comparator //insert comparator in the correct spot
      }
      gap = Math.floor(gap/2) //increment the gap
  }
  return array
}

var twoSum = function(nums, target) {
  let array = shell_sort(nums.slice())
  for (ind = 0; ind < array.length; ind ++) {
      for (i = array.length-1; i > ind; i--) {
          if (array[ind] + array[i] === target) {
              if (array[ind] != array[i]) {
                  return [nums.indexOf(array[i]), nums.indexOf(array[ind])]
              } else {
                  let foo = nums.indexOf(array[ind])  
                  return [foo, nums.slice(foo+1).indexOf((array[i]))+foo+1]
              }
          } s
      }
  }
}

File: easy/Two_Sum/Two_Sum.cpp
#include <vector>

// Runtime: 8 ms, faster than 93.28% of C++ online submissions for Two Sum.
// Memory Usage: 9.4 MB, less than 67.57% of C++ online submissions for Two Sum.
// 02/25/2020
// MM/DD/YYYY

int indexOf(std::vector<int>& list, int target) {
    for (unsigned int i = 0; i < list.size(); i++) {
        if (target == list[i]) {
            return i;
        }
    }
    return -1;
}

std::vector<int> shell_sort(std::vector<int> &array) {
    int gap = array.size()/2;
    while (gap >= 1) {
        for (int i = gap; i < array.size(); i++) { //iterate through array, starting from gap
            int comparator = array[i]; //make comparisons with this
            int output; //for accessing x outside the array
            int index; //for negative indexes

            for (int x = i; x > gap-2; x -= gap) { //iterate throguh array with gap as the step
                output = x; //to access x outside the loop
                if (x-gap < 0) { //in case of negative index
                    index = array.size()-x-gap;
                } else {
                    index = x-gap;
                }

                if (array[index] <= comparator) { //break when correct spot is found
					break;
				} else { //otherwise, move elements forward to make space
					array[x] = array[index];
				}
            }
            array[output] = comparator; //insert comparator in the correct spot
        }
        gap /= 2; //increment the gap
    }
    return array;
}


class Solution {
    public:
        std::vector<int> twoSum(std::vector<int>& nums, int target) {
            std::vector<int> input_array(nums);
            std::vector<int> array = shell_sort(input_array);
            
            for (unsigned int ind = 0; ind < array.size(); ind ++) {
                for (unsigned int i = array.size()-1; i > ind; i--) {
                    if (array[ind] + array[i] == target) {
                        if (array[ind] != array[i]) {
                            std::vector<int> output = {
                                indexOf(nums, array[i]), 
                                indexOf(nums, array[ind])
                            };
                            return output;
                        } else {
                            int foo = indexOf(nums, array[ind]);
                            std::vector<int> slice = std::vector<int>(nums.begin() + foo + 1, nums.end());
                            std::vector<int> output = {
                                foo,
                                indexOf(slice, (array[i]))+foo+1
                            };
                            return output;
                        }
                    } else if (target-array[ind] > array[i]) {
                        break;
                    }
                }
            }
            std::vector<int> out = {};
            return out;
        }
};